### PR TITLE
Chill CRUD

### PR DESCRIPTION
**Pull request by lhamu**

### CODE REVIEW

File: src/middlewares/checkToken.js [hunk 1]
```diff
@@ -3,37 +3,37 @@ import HttpStatus from 'http-status-codes';
 import * as tokenServices from '../services/token';
 
 const AUTHENTICATION_SALT_KEY = 'CHILL_RESTFULAPI';
-const REFRESHTOKEN_AUTHENTICATION_SALT_KEY='CHILL_REFRESH';
+const REFRESHTOKEN_AUTHENTICATION_SALT_KEY = 'CHILL_REFRESH';
```
Comment (romitamgai) [inline src/middlewares/checkToken.js]: lets keep these in env file

File: src/middlewares/checkToken.js [hunk 2]
```diff
@@ -3,37 +3,37 @@ import HttpStatus from 'http-status-codes';
...
-  } catch(err) {
-    return res.status(HttpStatus.UNAUTHORIZED).json({message: "Invalid authorization token."});
+  } catch (err) {
+    return res.status(HttpStatus.UNAUTHORIZED).json({ message: 'Invalid authorization token.' });
   }
   req.userId = tokenPayload.userId;
   next();
 }
 
 export async function authenticateRefreshToken(req, res, next) {
```
Comment (romitamgai) [inline src/middlewares/checkToken.js]: refactor this like that of authentication middleware

File: src/models/Service.js [hunk 3]
```diff
@@ -44,64 +44,70 @@ class Service extends db.Model {
...
-      ifMatch= await db.knex.raw(
-        userProjectQuery.CHECK_USER_PROJECT , [userId,id]
+
+    try {
+      ifMatch = await db.knex.raw(
+        userProjectQuery.CHECK_USER_PROJECT, [userId, id]
       );
     } catch (err) {
-    logger().error('Error while persisting the service into database', err);
+      logger().error('Error while persisting the service into database', err);
```
Comment (romitamgai) [inline src/models/Service.js]: don't we need to throw error here?

File: src/models/Service.js [hunk 4]
```diff
@@ -44,64 +44,70 @@ class Service extends db.Model {
...
-      ifMatch= await db.knex.raw(
-        userProjectQuery.CHECK_USER_PROJECT , [userId,projectId]
+
+    try {
+      ifMatch = await db.knex.raw(
+        userProjectQuery.CHECK_USER_PROJECT, [userId, projectId]
       );
     } catch (err) {
-    logger().error('Error while persisting the service into database', err);
+      logger().error('Error while persisting the service into database', err);
```
Comment (romitamgai) [inline src/models/Service.js]: do we need to throw err from here?

File: src/services/token.js [hunk 5]
```diff
@@ -2,8 +2,8 @@ import jwt from 'jsonwebtoken';
 import Token from '../models/Token';
 import logger from '../utils/logger';
 import * as generateTokens from '../jwt';
-const AUTHORIZATION_SALT_KEY='CHILL_RESTFULAPI';
-const REFRESH_TOKEN_SALT_KEY='CHILL_REFRESH';
+const AUTHORIZATION_SALT_KEY = 'CHILL_RESTFULAPI';
+const REFRESH_TOKEN_SALT_KEY = 'CHILL_REFRESH';
```
Comment (romitamgai) [inline src/services/token.js]: keep these in env file

File: src/services/tokenGeneration.js [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import * as generateTokens from '../jwt';
-const AUTHORIZATION_SALT_KEY='CHILL_RESTFULAPI';
-const REFRESH_TOKEN_SALT_KEY='CHILL_REFRESH';
+const AUTHORIZATION_SALT_KEY = 'CHILL_RESTFULAPI';
+const REFRESH_TOKEN_SALT_KEY = 'CHILL_REFRESH';
```
Comment (romitamgai) [inline src/services/tokenGeneration.js]: keep these in env file

File: src/services/user.js [hunk 7]
```diff
@@ -4,14 +4,10 @@ import User from '../models/User';
 import logger from '../utils/logger';
 import * as tokenService from './token';
 import * as generateTokens from '../jwt';
-<<<<<<< HEAD
+
 const AUTHORIZATION_SALT_KEY = 'CHILL_RESTFULAPI';
-const REFRESHTOKEN_SALT_KEY = 'CHILL_REFRESH';
+const REFRESH_TOKEN_SALT_KEY = 'CHILL_REFRESH';
```
Comment (romitamgai) [inline src/services/user.js]: keep these in env file

**Review by romitamgai — CHANGES_REQUESTED** [review]

File: src/queries/userProject.js [hunk 8]
```diff
@@ -0,0 +1,6 @@
+export const CHECK_USER_PROJECT = (`
```
Comment (kabirbaidhya) [inline src/queries/userProject.js]:
```js
export const CHECK_USER_PROJECT = `
  SELECT first_name, last_name
  FROM table_name
  INNER JOIN test ON t.id = s.user_id
  WHERE id = :userId
  GROUP BY
`;

File: src/services/service.js [hunk 9]
```diff
@@ -8,8 +8,8 @@ import StatusLog from '../models/StatusLog';
  *
  * @return {Promise}
  */
-export async function fetchAll(id) {
-  return Service.fetchAll(id);
+export async function fetchAll( id ,  userId ) {
```
Comment (kabirbaidhya) [inline src/services/service.js]:
```
export async function fetchAll(id, userId) {
```

Comment (kabirbaidhya) [reply src/services/service.js]:
```
let a = { name: 'test' };

import { name } from 'test';
```

Comment (mesaugat) [reply src/services/service.js]: We need prettier.

**Comment by kabirbaidhya** [comment]
Install `prettier` extenstion in `vscode`. Enable it on save after that setting `"editor.formatOnSave": true`. @lhamu @andreshsingh @leapfrogtechnology/the-chill-team

File: src/queries/service.js [hunk 10]
```diff
@@ -1,27 +1,26 @@
 export const FETCH_All_SERVICE = (`
```
Comment (kabirbaidhya) [inline src/queries/service.js]: let's remove the additional bracket `(` or `)` from all the raw queries.

File: src/queries/project.js [hunk 11]
```diff
@@ -1,38 +1,34 @@
...
+    DELETE
+    FROM projects
+    WHERE id=?
     `);
 
 export const DELETE_A_PROJECT_USERPROJECT = (`
-    Delete
-    from user_project
-    where project_id=?
+    DELETE
```
Comment (kabirbaidhya) [inline src/queries/project.js]: Use raw queries ONLY for complex `SELECT` queries that affect readability if done with query builder. Not for `UPDATE`, `DELETE`, `INSERT` and trivial queries like `SELECT * FROM table`.

**Review by kabirbaidhya — CHANGES_REQUESTED** [review]

File: src/middlewares/checkToken.js [hunk 12]
```diff
@@ -1,43 +1,45 @@
...
 
-/*
-to authenticate the access token
-*/
-export async function authenticate( req, res, next) {
-  let token = req.headers.authorization;
-  let head = req.headers.authorization.split(' ')[1];
-  let value = jwt.verify(head, 'RESTFULAPI');
+const AUTHENTICATION_SALT_KEY = 'CHILL_RESTFULAPI';
+const REFRESHTOKEN_AUTHENTICATION_SALT_KEY = 'CHILL_REFRESH';
```
Comment (pratishshr) [inline src/middlewares/checkToken.js]: Get the salt from chill.yml
Salts should not be in the code itself.

File: src/config/auth.js [hunk 13]
```diff
@@ -1,7 +1,8 @@
 module.exports = {
-  'googleAuth': {
-    'clientID': '725194704366-4nlruh4oc7g2lrtfok18qqacnq7h2o2m.apps.googleusercontent.com',
-    'clientSecret': 'xjJrXQs2KRT3NBqwXPBue5sy',
-    'callbackURL': 'http://localhost:8000/api/auth/google/callback'
+  googleAuth: {
```
Comment (kabirbaidhya) [inline src/config/auth.js]: We store all our configuration in the `chill.yml` file. No need additional config files.

File: .gitignore [hunk 14]
```diff
@@ -35,3 +35,9 @@ coverage
 # Chill database
 chill.db
 chill.yml
+
+# Google OAuth details
+src/config/auth.js
```
Comment (kabirbaidhya) [inline .gitignore]: No need to ignore additional config files. As we already have `chill.yml`

File: src/app.js [hunk 15]
```diff
@@ -16,7 +16,6 @@ const app = express();
 import passport from 'passport';
 import passportConfig from './config/passport';
 
-
 passportConfig(passport);
```
Comment (kabirbaidhya) [inline src/app.js]:
```js
passportConfig(config.get().passport);
```

File: src/app.js [hunk 16]
```diff
@@ -16,7 +16,6 @@ const app = express();
 import passport from 'passport';
 import passportConfig from './config/passport';
```
Comment (kabirbaidhya) [inline src/app.js]: No need to import additional config files

File: src/queries/project.js [hunk 17]
```diff
@@ -1,38 +1,34 @@
...
-    left join user_project up on up.user_id = u.id 
-    left join projects p on up.project_id = p.id 
-    where u.id = ? ;
-    `);
+export const FETCH_All_PROJECTS = `
+  SELECT * 
+  FROM users u 
+  LEFT JOIN user_project up on up.user_id = u.id 
+  LEFT JOIN projects p on up.project_id = p.id 
+  WHERE u.id = ? ;
```
Comment (kabirbaidhya) [inline src/queries/project.js]: named parameter

Comment (kabirbaidhya) [inline src/queries/project.js]: named parameter

File: src/queries/project.js [hunk 18]
```diff
@@ -1,38 +1,34 @@
...
+  FROM projects
+  WHERE id=?
+  `;
 
-export const DELETE_A_PROJECT_USERPROJECT = (`
-    Delete
-    from user_project
-    where project_id=?
-    `)
+export const DELETE_A_PROJECT_USERPROJECT = `
```
Comment (kabirbaidhya) [inline src/queries/project.js]: here

File: src/queries/project.js [hunk 19]
```diff
@@ -1,38 +1,34 @@
...
+  WHERE project_id=?
+  `;
 
-export const UPDATE_A_PROJECT_PROJECTS = (`
-    update projects
-    set name = ? , description = ?
-    where id = ?
-    `)
\ No newline at end of file
+export const UPDATE_A_PROJECT_PROJECTS = `
```
Comment (kabirbaidhya) [inline src/queries/project.js]: here

File: src/queries/service.js [hunk 20]
```diff
@@ -0,0 +1,26 @@
+export const FETCH_All_SERVICE = `
+  SELECT *
```
Comment (kabirbaidhya) [inline src/queries/service.js]: no need raw query for this

File: src/queries/service.js [hunk 21]
```diff
@@ -0,0 +1,26 @@
+export const FETCH_All_SERVICE = `
+  SELECT * 
+  FROM services
+  WHERE project_id = ? ;
+  `;
+
+export const FETCH_A_SERVICE = `
```
Comment (kabirbaidhya) [inline src/queries/service.js]: and this

File: src/queries/service.js [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+  SELECT * 
+  FROM services
+  WHERE id = ?
+  AND project_id = ?;
+  `;
+
+export const DELETE_A_SERVICE = `
```
Comment (kabirbaidhya) [inline src/queries/service.js]: this as well

File: src/queries/service.js [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+  DELETE 
+  FROM services
+  WHERE id = ?
+  AND project_id = ?;
+  `;
+
+export const UPDATE_A_SERVICE_SERVICES = `
```
Comment (kabirbaidhya) [inline src/queries/service.js]: this

File: src/queries/userProject.js [hunk 24]
```diff
@@ -0,0 +1,6 @@
+export const CHECK_USER_PROJECT = `
+	SELECT *
```
Comment (kabirbaidhya) [inline src/queries/userProject.js]: indentation

File: src/services/service.js [hunk 25]
```diff
@@ -4,38 +4,32 @@ import Service from '../models/Service';
...
-  return result;
+export async function fetchAll(id, userId) {
+  return Service.fetchAll(id, userId);
 }
 
 /**
  * Fetch a single service by it's id (pk).
  *
- * @param  {string|Number}  id
+ * @param  {string|Number}  projectId, serviceId
```
Comment (kabirbaidhya) [inline src/services/service.js]: incorrect doc block

File: src/services/service.js [hunk 26]
```diff
@@ -72,3 +68,12 @@ export async function create(data) {
   }
 }
 
+/**
+ * Update and save service data.
+ *
+ * @param {Object}
```
Comment (kabirbaidhya) [inline src/services/service.js]: incorrect

File: src/services/token.js [hunk 27]
```diff
@@ -1,68 +1,55 @@
+import jwt from 'jsonwebtoken';
 import Token from '../models/Token';
 import logger from '../utils/logger';
+import * as key from '../config/key';
```
Comment (kabirbaidhya) [inline src/services/token.js]: here

File: src/services/token.js [hunk 28]
```diff
@@ -1,68 +1,55 @@
...
- * Function to generate new access token when it expires, after validating the refresh token
- * 
- */
 export async function generateAccessToken(refresh_token) {
   try {
     let result = await fetchToken(refresh_token);
     let refreshToken = result.refresh_token;
 
-    jwt.verify(refreshToken, 'REFRESH');
+    jwt.verify(refreshToken, key.REFRESH_TOKEN_SALT_KEY);
```
Comment (kabirbaidhya) [inline src/services/token.js]:
```js
jwt.verify(refreshToken, config.get().auth.saltKey);
```

File: src/services/token.js [hunk 29]
```diff
@@ -1,68 +1,55 @@
...
     if (result) {
-      let accessToken = generateTokens.generateToken(result.user_id, 'RESTFULAPI', 300);
+      let accessToken = generateTokens.generateToken(
+        result.user_id,
+        key.AUTHORIZATION_SALT_KEY,
```
Comment (kabirbaidhya) [inline src/services/token.js]:
```js
config.get().auth.saltKey
```

File: src/services/tokenGeneration.js [hunk 30]
```diff
@@ -1,16 +1,20 @@
...
+  return new Promise(resolve => {
+    let accessToken = generateTokens.generateToken(
+      id,
+      key.AUTHORIZATION_SALT_KEY,
+      300
+    );
+
+    let refreshToken = generateTokens.generateToken(
+      id,
+      key.REFRESH_TOKEN_SALT_KEY,
```
Comment (kabirbaidhya) [inline src/services/tokenGeneration.js]:
```
config.get().auth.saltKey
```

File: src/services/user.js [hunk 31]
```diff
@@ -31,39 +30,62 @@ export async function createUser(data) {
...
     let email = data.email;
     let userInfo = await fetchByEmail(email);
 
-    console.log(userInfo);
-
     if (userInfo) {
-      let accessToken = generateTokens.generateToken(userInfo.id, 'RESTFULAPI', 300);
+      let accessToken = generateTokens.generateToken(
+        userInfo.id,
+        key.AUTHORIZATION_SALT_KEY,
```
Comment (kabirbaidhya) [inline src/services/user.js]:
```js
config.get().auth.saltKey
```

File: src/services/user.js [hunk 32]
```diff
@@ -31,39 +30,62 @@ export async function createUser(data) {
...
+        return { accessToken, refreshToken };
+      }
+    } else {
+      await createUser(data);
       let userInfo = await fetchByEmail(data.email);
-      let accessToken = generateTokens.generateToken(userInfo.id, 'RESTFULAPI', 300);
-      let refreshToken = generateTokens.generateToken(userInfo.id, 'REFRESH', 172800);
+      let accessToken = generateTokens.generateToken(
+        userInfo.id,
+        key.AUTHORIZATION_SALT_KEY,
```
Comment (kabirbaidhya) [inline src/services/user.js]: here

File: src/services/user.js [hunk 33]
```diff
@@ -31,39 +30,62 @@ export async function createUser(data) {
...
+        300
+      );
+      let refreshToken = generateTokens.generateToken(
+        userInfo.id,
+        key.REFRESH_TOKEN_SALT_KEY,
```
Comment (kabirbaidhya) [inline src/services/user.js]: here

**Review by kabirbaidhya — CHANGES_REQUESTED** [review]

**Comment by kabirbaidhya** [comment]
Reminder: As previously added by @deepsadhi in the other PR

1. Import npm packages at top and custom packages separated by blank line
2. Indent code properly with spacing  
3. Use of doc block 
4. Remove use of return keyword if return statement is of single line `[  func1(obj => func2(obj2))  ]` 
5. Remove console.log before code is pushed
6. Remove ' from object key. Eg. `{ name: 'ABC' }` instead of `{ 'name': 'ABC' }`
7. If key and variable is same of object use `{ name }` instead of `{ name: name }`
8. Order statements by statement length order
9. Follow same if syntax indentation by everyone
10. Remove comments if not required
11. Do not leave multiple lines, one line is enough
12. Write all SQL keywords in UPPERCASE
13. Remove magic numbers. Use constant instead of it
14. Use of single quote ' instead of double quote "

@lhamu @andreshsingh

**Review by romitamgai — APPROVED** [review]

**Pull request merged**